tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -1586,36 +1591,49 @@ class ReduceConverter : public OpConversionPattern<lmhlo::ReduceOp> {
       // The incoming region is operating on buffers, while linalg.generic
       // expects scalar SSA values. Add some allocs around the original op to
       // make it compatible.
-      auto arg_type = block->getArgument(0).getType().cast<MemRefType>();
-      Value alloc_a = rewriter.create<memref::AllocaOp>(loc, arg_type);
-      Value alloc_b = rewriter.create<memref::AllocaOp>(loc, arg_type);
-      Value alloc_res = rewriter.create<memref::AllocaOp>(loc, arg_type);
+      SmallVector<MemRefType, 4> mem_argv_tys;
+      SmallVector<Value, 4> alloc_values;
+      for (size_t i = 0; i < block->getNumArguments(); ++i) {
+        mem_argv_tys.push_back(block->getArgumentTypes()[i].cast<MemRefType>());
+        alloc_values.push_back(
+            rewriter.create<memref::AllocaOp>(loc, mem_argv_tys.back()));
+      }
+      size_t input_num = alloc_values.size() / 3 * 2;
 
       // Now turn the existing signature
       //   (memref<X>, memref<X>, memref<X>) -> ()
       // into
       //   (X, X) -> X
-      TypeConverter::SignatureConversion signature_converter(3);
-      signature_converter.remapInput(0, alloc_a);
-      signature_converter.remapInput(1, alloc_b);
-      signature_converter.remapInput(2, alloc_res);
-      signature_converter.addInputs(
-          {arg_type.getElementType(), arg_type.getElementType()});
+      TypeConverter::SignatureConversion signature_converter(
+          alloc_values.size());
+      for (auto it : llvm::enumerate(alloc_values)) {
+        signature_converter.remapInput(it.index(), it.value());
+      }
+      SmallVector<Type, 4> element_tys;
+      for (size_t i = 0; i < input_num; ++i)
+        element_tys.push_back(mem_argv_tys[i].getElementType());
```

Comment:
style nit: Wrap with braces if it's not single line statement.

Also, you could iterate it without indices. I.e.

```cpp
for (auto ty : mem_argv_tys) {
  ...
}
```